File: backend/app/agent.py
"""CareSync Strands Agent Core.

Orchestrates medication extraction, history lookup, drug conflict checking,
and daily chronotherapy schedule generation using the Strands Agents SDK.
"""

import json
import logging
import os
import re
import urllib.request
from typing import Any, Dict, List, Optional

from strands import Agent, tool
try:
    from app.tools import (
        fetch_patient_history,
        check_drug_interaction,
        generate_daily_schedule,
    )
except ImportError:
    from tools import (
        fetch_patient_history,
        check_drug_interaction,
        generate_daily_schedule,
    )

logger = logging.getLogger("caresync_agent")
logger.setLevel(logging.INFO)

OLLAMA_BASE_URL = os.environ.get("OLLAMA_BASE_URL", "http://localhost:11434")
OLLAMA_MODEL = os.environ.get("OLLAMA_MODEL", "llama3")

CARESYNC_SYSTEM_PROMPT = """You are CareSync, an expert geriatric clinical AI assistant for elderly care.
Your duty is to assist families by processing unstructured doctor dictation notes, cross-referencing
active medications for the patient, catching adverse drug-drug interactions, and formatting
a clean, safe daily pill schedule.

You MUST use your registered tools:
1. fetch_patient_history(patient_id): Look up current medications on file.
2. check_drug_interaction(new_meds, current_meds): Evaluate potential adverse reactions.
3. generate_daily_schedule(medications): Organize pills into morning, afternoon, and evening slots.

Always emphasize safety and flag any severe drug conflicts.
"""


def _is_ollama_reachable(base_url: str = OLLAMA_BASE_URL) -> bool:
    """Check if the local Ollama instance is online."""
    try:
        req = urllib.request.Request(f"{base_url}/api/tags", method="GET")
        with urllib.request.urlopen(req, timeout=1.5) as resp:
            return resp.status == 200
    except Exception:
        return False


def _query_ollama_llm(prompt: str, system_prompt: str = CARESYNC_SYSTEM_PROMPT) -> Optional[str]:
    """Query local Ollama instance on localhost:11434."""
    if not _is_ollama_reachable():
        return None
    try:
        payload = json.dumps({
            "model": OLLAMA_MODEL,
            "prompt": prompt,
            "system": system_prompt,
            "stream": False
        }).encode("utf-8")
        req = urllib.request.Request(
            f"{OLLAMA_BASE_URL}/api/generate",
            data=payload,
            headers={"Content-Type": "application/json"},
            method="POST"
        )
        with urllib.request.urlopen(req, timeout=10.0) as resp:
            data = json.loads(resp.read().decode("utf-8"))
            return data.get("response", "")
    except Exception as e:
        logger.warning(f"Ollama query failed: {e}")
        return None


def extract_medications_from_note(doctors_note: str) -> List[Dict[str, Any]]:
    """Extract medication details from unstructured clinical text using regex patterns.

    Handles doctor notes like:
    'Hi, this is Dr. Smith. I need Grandma Bob to start taking 400mg of Ibuprofen
     twice a day for her knee pain, morning and evening. She should continue her other meds.'
    """
    meds_found = []

    # Common medications in eldercare to look for
    drug_catalog = [
        "ibuprofen", "lisinopril", "warfarin", "aspirin", "metformin",
        "atorvastatin", "amlodipine", "omeprazole", "levothyroxine",
        "tramadol", "sertraline", "gabapentin", "furosemide"
    ]

    note_lower = doctors_note.lower()

    for drug in drug_catalog:
        if drug in note_lower:
            # Extract dosage (e.g. 400mg, 10mg)
            dosage_match = re.search(rf"{drug}\s+(\d+\s*(?:mg|mcg|g|ml))", note_lower)
            if not dosage_match:
                dosage_match = re.search(rf"(\d+\s*(?:mg|mcg|g|ml))\s+(?:of\s+)?{drug}", note_lower)
            dosage = dosage_match.group(1).replace(" ", "") if dosage_match else "Standard dose"

            # Extract frequency & timing
            frequency = "as needed"
            timing = []
            if "twice a day" in note_lower or "2x daily" in note_lower or "bid" in note_lower:
                frequency = "twice a day"
            elif "once a day" in note_lower or "daily" in note_lower or "qd" in note_lower:
                frequency = "once a day"
            elif "three times a day" in note_lower or "tid" in note_lower:
                frequency = "three times a day"

            if "morning" in note_lower:
                timing.append("morning")
            if "afternoon" in note_lower:
                timing.append("afternoon")
            if "evening" in note_lower or "night" in note_lower:
                timing.append("evening")

            # Extract indication
            reason_match = re.search(r"for (?:her|his|their)?\s*([a-zA-Z\s]+?)(?:,|\.|\bmorning\b|\bevening\b)", note_lower)
            reason = reason_match.group(1).strip() if reason_match else "prescribed symptom relief"

            med_obj = {
                "name": f"{drug.capitalize()} {dosage}".strip(),
                "drug": drug.capitalize(),
                "dosage": dosage,
                "frequency": frequency,
                "timing": timing if timing else ["morning"],
                "reason": reason
            }
            meds_found.append(med_obj)

    # Fallback if no specific catalog match found
    if not meds_found:
        generic_match = re.findall(r"(\d+\s*mg)\s+(?:of\s+)?([A-Za-z]+)", doctors_note, re.IGNORECASE)
        for dose, name in generic_match:
            meds_found.append({
                "name": f"{name.capitalize()} {dose}".strip(),
                "drug": name.capitalize(),
                "dosage": dose,
                "frequency": "as directed",
                "timing": ["morning"],
                "reason": "clinical indication"
            })

    return meds_found


class CareSyncStrandsAgent:
    """CareSync Strands Agent wrapper managing execution and tool invocations."""

    def __init__(self, ollama_url: str = OLLAMA_BASE_URL):
        self.ollama_url = ollama_url
        self.tools = [
            fetch_patient_history,
            check_drug_interaction,
            generate_daily_schedule,
        ]
        self.system_prompt = CARESYNC_SYSTEM_PROMPT
        self._init_strands_agent()

    def _init_strands_agent(self):
        """Initialize underlying Strands SDK Agent instance."""
        try:
            self.agent = Agent(
                tools=self.tools,
                system_prompt=self.system_prompt,
                name="CareSyncClinicalAgent",
                description="Autonomous medication orchestrator and drug conflict detector"
            )
        except Exception as e:
            logger.warning(f"Strands Agent init with default config: {e}")
            self.agent = None

    def run_golden_path(self, patient_id: str, doctors_note: str) -> Dict[str, Any]:
        """Execute the complete Strands Agent execution loop as specified in the blueprint.

        Agent Loop:
        1. Agent uses fetch_patient_history(patient_id).
        2. Agent parses new meds from doctors_note.
        3. Agent uses check_drug_interaction(new_meds, current_meds).
        4. If conflict, Agent notes the warning and severity.
        5. Agent uses generate_daily_schedule(medications).
        6. Returns structured payload.
        """
        logger.info(f"Starting CareSync Strands Agent loop for patient: {patient_id}")

        # Step 1: Agent uses fetch_patient_history
        history_result = fetch_patient_history(patient_id=patient_id)
        current_meds = history_result.get("current_medications", [])
        patient_name = history_result.get("name", patient_id)

        # Step 2: Agent parses the new meds from note (via local LLM if online, else deterministic extractor)
        llm_output = _query_ollama_llm(
            prompt=f"Patient: {patient_name}. Current medications: {current_meds}. Doctor's note: {doctors_note}. Extract newly prescribed medications and instructions."
        )
        new_meds_parsed = extract_medications_from_note(doctors_note)

        # Format string lists for tool execution
        new_meds_names = [m["name"] for m in new_meds_parsed]

        # Step 3: Agent uses check_drug_interaction
        interaction_result = check_drug_interaction(
            new_meds=new_meds_names,
            current_meds=current_meds
        )

        interaction_warnings = []
        if interaction_result.get("conflict_found"):
            conflicts = interaction_result.get("conflicts", [])
            if conflicts:
                for c in conflicts:
                    interaction_warnings.append({
                        "drugs": [c["pair"].split(" + ")[0], c["pair"].split(" + ")[1]] if " + " in c.get("pair", "") else [interaction_result.get("details", "")],
                        "severity": c.get("severity", "High"),
                        "warning": c.get("warning", ""),
                        "clinical_guidance": c.get("clinical_guidance", "Exercise clinical caution.")
                    })
            else:
                interaction_warnings.append({
                    "drugs": [current_meds[0] if current_meds else "Current", new_meds_names[0] if new_meds_names else "New"],
                    "severity": interaction_result.get("severity", "High"),
                    "warning": interaction_result.get("warning", interaction_result.get("details", "")),
                    "clinical_guidance": "Consult prescribing clinician immediately."
                })

        # Step 4: Agent uses generate_daily_schedule
        # Combine current medications with newly prescribed medications
        combined_meds_for_schedule = []
        for cm in current_meds:
            combined_meds_for_schedule.append({
                "medication": cm,
                "name": cm,
                "timing": ["morning"],
                "instructions": "Existing maintenance regimen. Take with water."
            })
        for nm in new_meds_parsed:
            combined_meds_for_schedule.append({
                "medication": nm["name"],
                "name": nm["name"],
                "timing": nm.get("timing", ["morning"]),
                "instructions": f"New prescription for {nm.get('reason', 'symptom relief')}. {nm.get('frequency', '')}."
            })

        schedule_result = generate_daily_schedule(medications=combined_meds_for_schedule)

        # Assemble final structured response
        response_payload = {
            "status": "success",
            "patient_id": patient_id,
            "patient_name": patient_name,
            "doctor_note_processed": doctors_note,
            "llm_engine": f"Ollama ({OLLAMA_MODEL})" if _is_ollama_reachable() else "Strands Deterministic Engine (Offline Fallback)",
            "current_medications": current_meds,
            "new_medications_detected": new_meds_parsed,
            "conflict_found": interaction_result.get("conflict_found", False),
            "interaction_warnings": interaction_warnings,
            "daily_schedule": schedule_result,
            "agent_execution_summary": {
                "tools_invoked": [
                    "fetch_patient_history",
                    "check_drug_interaction",
                    "generate_daily_schedule"
                ],
                "conflicts_count": len(interaction_warnings)
            }
        }

        logger.info(f"Agent loop complete. Conflicts detected: {len(interaction_warnings)}")
        return response_payload


# Singleton agent instance
default_agent = CareSyncStrandsAgent()

File: backend/src/agent.ts
/**
 * CareSync Strands Agent Core in TypeScript.
 * 
 * Orchestrates medication extraction, history lookup, drug conflict checking,
 * and daily chronotherapy schedule generation.
 */

import { fetchPatientHistory, checkDrugInteraction, generateDailySchedule } from './tools.js';
import type { AgentResult, ExtractedMedication, InteractionWarning } from './types.js';

export const CARESYNC_SYSTEM_PROMPT = `You are CareSync, an expert geriatric clinical AI assistant for elderly care.
Your duty is to assist families by processing unstructured doctor dictation notes, cross-referencing
active medications for the patient, catching adverse drug-drug interactions, and formatting
a clean, safe daily pill schedule.

You MUST use your registered tools:
1. fetch_patient_history(patient_id): Look up current medications on file.
2. check_drug_interaction(new_meds, current_meds): Evaluate potential adverse reactions.
3. generate_daily_schedule(medications): Organize pills into morning, afternoon, and evening slots.

Always emphasize safety and flag any severe drug conflicts.`;

/**
 * Extracts medication details from clinical text using regex and heuristics.
 */
export function extractMedicationsFromNote(doctorsNote: string): ExtractedMedication[] {
  const medsFound: ExtractedMedication[] = [];

  const drugCatalog = [
    'ibuprofen',
    'lisinopril',
    'warfarin',
    'aspirin',
    'metformin',
    'atorvastatin',
    'amlodipine',
    'omeprazole',
    'levothyroxine',
    'tramadol',
    'sertraline',
    'gabapentin',
    'furosemide',
  ];

  const noteLower = doctorsNote.toLowerCase();

  for (const drug of drugCatalog) {
    if (noteLower.includes(drug)) {
      // Extract dosage
      let dosage = 'Standard dose';
      const dosageMatch = noteLower.match(new RegExp(`${drug}\\s+(\\d+\\s*(?:mg|mcg|g|ml))`, 'i')) ||
        noteLower.match(new RegExp(`(\\d+\\s*(?:mg|mcg|g|ml))\\s+(?:of\\s+)?${drug}`, 'i'));
      if (dosageMatch) {
        dosage = dosageMatch[1].replace(/\s+/g, '');
      }

      // Extract frequency & timing
      let frequency = 'as needed';
      const timing: string[] = [];

      if (noteLower.includes('twice a day') || noteLower.includes('2x daily') || noteLower.includes('bid')) {
        frequency = 'twice a day';
      } else if (noteLower.includes('once a day') || noteLower.includes('daily') || noteLower.includes('qd')) {
        frequency = 'once a day';
      } else if (noteLower.includes('three times a day') || noteLower.includes('tid')) {
        frequency = 'three times a day';
      }

      if (noteLower.includes('morning')) timing.push('morning');
      if (noteLower.includes('afternoon')) timing.push('afternoon');
      if (noteLower.includes('evening') || noteLower.includes('night')) timing.push('evening');

      // Extract indication/reason
      const reasonMatch = noteLower.match(/for (?:her|his|their)?\s*([a-zA-Z\s]+?)(?:,|\.|\bmorning\b|\bevening\b)/i);
      const reason = reasonMatch ? reasonMatch[1].trim() : 'prescribed symptom relief';

      const capitalizedDrug = drug.charAt(0).toUpperCase() + drug.slice(1);
      medsFound.push({
        name: `${capitalizedDrug} ${dosage}`.trim(),
        drug: capitalizedDrug,
        dosage,
        frequency,
        timing: timing.length > 0 ? timing : ['morning'],
        reason,
      });
    }
  }

  // Fallback if no catalog matches
  if (medsFound.length === 0) {
    const genericMatch = doctorsNote.matchAll(/(\d+\s*mg)\s+(?:of\s+)?([A-Za-z]+)/gi);
    for (const match of genericMatch) {
      const dose = match[1];
      const name = match[2].charAt(0).toUpperCase() + match[2].slice(1).toLowerCase();
      medsFound.push({
        name: `${name} ${dose}`.trim(),
        drug: name,
        dosage: dose,
        frequency: 'as directed',
        timing: ['morning'],
        reason: 'clinical indication',
      });
    }
  }

  return medsFound;
}

export class CareSyncStrandsAgent {
  /**
   * Execute the complete Strands Agent execution loop.
   */
  public runGoldenPath(patientId: string, doctorsNote: string): AgentResult {
    // Step 1: Agent uses fetch_patient_history
    const historyResult = fetchPatientHistory(patientId);
    const currentMeds = historyResult.current_medications || [];
    const patientName = historyResult.name || patientId;

    // Step 2: Agent parses the new meds from note
    const newMedsParsed = extractMedicationsFromNote(doctorsNote);
    const newMedsNames = newMedsParsed.map((m) => m.name);

    // Step 3: Agent uses check_drug_interaction
    const interactionResult = checkDrugInteraction(newMedsNames, currentMeds);

    const interactionWarnings: InteractionWarning[] = [];
    if (interactionResult.conflict_found) {
      if (interactionResult.conflicts && interactionResult.conflicts.length > 0) {
        for (const c of interactionResult.conflicts) {
          interactionWarnings.push({
            drugs: c.drugs,
            severity: c.severity || 'High',
            warning: c.warning,
            clinical_guidance: c.clinical_guidance || 'Exercise clinical caution.',
          });
        }
      } else {
        interactionWarnings.push({
          drugs: [currentMeds[0] || 'Current', newMedsNames[0] || 'New'],
          severity: interactionResult.severity || 'High',
          warning: interactionResult.warning || interactionResult.details,
          clinical_guidance: 'Consult prescribing clinician immediately.',
        });
      }
    }

    // Step 4: Agent uses generate_daily_schedule
    const combinedMedsForSchedule: Array<{ medication: string; name: string; timing: string[]; instructions: string }> = [];

    for (const cm of currentMeds) {
      combinedMedsForSchedule.push({
        medication: cm,
        name: cm,
        timing: ['morning'],
        instructions: 'Existing maintenance regimen. Take with water.',
      });
    }

    for (const nm of newMedsParsed) {
      combinedMedsForSchedule.push({
        medication: nm.name,
        name: nm.name,
        timing: nm.timing || ['morning'],
        instructions: `New prescription for ${nm.reason}. ${nm.frequency}.`,
      });
    }

    const scheduleResult = generateDailySchedule(combinedMedsForSchedule);

    return {
      status: 'success',
      patient_id: patientId,
      patient_name: patientName,
      doctor_note_processed: doctorsNote,
      llm_engine: 'Strands Agent Runtime (TypeScript Edition)',
      current_medications: currentMeds,
      new_medications_detected: newMedsParsed,
      conflict_found: interactionResult.conflict_found,
      interaction_warnings: interactionWarnings,
      daily_schedule: scheduleResult,
      agent_execution_summary: {
        tools_invoked: [
          'fetch_patient_history',
          'check_drug_interaction',
          'generate_daily_schedule',
        ],
        conflicts_count: interactionWarnings.length,
      },
    };
  }
}

export const defaultAgent = new CareSyncStrandsAgent();
